vulcan/lib/commands.rb

```diff
@@ -18,6 +19,174 @@ def setup(config)
     end
   end
 
+  class Workflow
+    include Etna::CommandExecutor
+
+    class Run
+      def execute(project_name, workflow_name)
+        @project_name = project_name
+        @workflow_name = workflow_name
+
+        while true
+          orchestration.run_until_done!(storage)
+          find_all_output_ui_queries.each do |query_name, source, path|
+            if self.respond_to?(query_name.to_sym)
+              self.send(query_name.to_sym, source, path)
+            else
+              show_data(source, path)
+            end
+          end
+
+          while (input = find_next_primary_input)
+            val = collect_primary_input(input.id, input.type)
+            session.define_user_input([:primary_inputs, input.id], val)
+          end
+
+          while (ui_query = find_next_input_ui_query)
```

Comment:
And these are actual ui-query steps that can use the run to encode a specific 'component'.  Like before, I use the query name to dynamically dispatch handling how to query data from the json, but otherwise fall back to a generic "give me a json input" workflow.